File: src/codes/ref-synergy/py/strategy-template.py
from abc import ABC, abstractmethod

class NotificationStrategy(ABC):
    @abstractmethod
    def send_message(self, to: str, content: str) -> None:
        pass

class EmailStrategy(NotificationStrategy):
    def send_message(self, to: str, content: str) -> None:
        print(f"[Email] To: {to}, Message: {content}")

class SlackStrategy(NotificationStrategy):
    def send_message(self, to: str, content: str) -> None:
        print(f"[Slack] To: {to}, Message: {content}")

class Notifier(ABC):
    def __init__(self, strategy: NotificationStrategy) -> None:
        self.strategy = strategy

    def send(self, user: str, message: str) -> None:
        print("--- Sending Start ---")
        content = self.format_message(user, message)
        self.strategy.send_message(user, content)
        print("--- Sending End ---")

    @abstractmethod
    def format_message(self, user: str, message: str) -> str:
        pass

class AlertNotifier(Notifier):
    def format_message(self, user: str, message: str) -> str:
        return f"[ALERT] {user}: {message}"

# 使用例
notifier = AlertNotifier(EmailStrategy())
notifier.send("taro@example.com", "サーバー障害が発生しました")

File: src/codes/ref-synergy/ts/strategy-template.ts
interface NotificationStrategy {
  sendMessage(to: string, content: string): void;
}

class EmailStrategy implements NotificationStrategy {
  sendMessage(to: string, content: string): void {
    console.log(`[Email] To: ${to}, Message: ${content}`);
  }
}

class SlackStrategy implements NotificationStrategy {
  sendMessage(to: string, content: string): void {
    console.log(`[Slack] To: ${to}, Message: ${content}`);
  }
}

abstract class Notifier {
  constructor(private strategy: NotificationStrategy) {}

  send(user: string, message: string): void {
    console.log("--- Sending Start ---");
    const content = this.formatMessage(user, message);
    this.strategy.sendMessage(user, content);
    console.log("--- Sending End ---");
  }

  protected abstract formatMessage(user: string, message: string): string;
}

class AlertNotifier extends Notifier {
  protected formatMessage(user: string, message: string): string {
    return `[ALERT] ${user}: ${message}`;
  }
}

// 使用例
const notifier = new AlertNotifier(new EmailStrategy());
notifier.send("taro@example.com", "サーバー障害が発生しました");
